File: ALM-Model-Lab/src/math/rates/simulateHw.ts
/**
 * Hull-White 1F Monte Carlo simulator — TypeScript port of `research/simulate_hw.py`.
 *
 * Exact transition for X under shifted-Gaussian HW with antithetic pairing
 * and multiplicative martingale correction (Glasserman Sec 4.5).
 */

import type { ZeroCurve } from "./bootstrap";
import { PCG32 } from "./random";
import { bFunction } from "./hwPricing";

export interface HWSimulationResult {
  seed: bigint;
  a: number;
  sigma: number;
  nPaths: number;
  horizonYears: number;
  dtYears: number;
  times: Float64Array;
  /** Latent factor X(t) per path per step. Shape: [nPaths][nSteps]. */
  XPaths: Float64Array[];
  /** Short rate r(t) per path per step. Shape: [nPaths][nSteps]. */
  rPaths: Float64Array[];
  dfMarket: Float64Array;
  dfSimulated: Float64Array;
  rawErrorBps: Float64Array;
  corrected: boolean;
}

export interface HWSimulationOptions {
  horizonYears?: number;
  dtYears?: number;
  nPairs?: number;
  seed?: bigint | number;
  applyCorrection?: boolean;
}

export function simulateHW(
  curve: ZeroCurve,
  a: number,
  sigma: number,
  options: HWSimulationOptions = {},
): HWSimulationResult {
  const horizonYears = options.horizonYears ?? 30.0;
  const dtYears = options.dtYears ?? 1 / 12;
  const nPairs = options.nPairs ?? 250;
  const seed = options.seed ?? 20250930n;
  const applyCorrection = options.applyCorrection ?? true;

  const rng = new PCG32(typeof seed === "bigint" ? seed : BigInt(seed));
  const nSteps = Math.round(horizonYears / dtYears);
  const nPaths = 2 * nPairs;
  const times = new Float64Array(nSteps);
  for (let k = 0; k < nSteps; k++) times[k] = (k + 1) * dtYears;

  // Precompute initial-curve forward rates at each step f^M(0, t_k).
  // Use f(t) = -log(P(0, t+dt)/P(0, t)) / dt evaluated at the step grid.
  const dfFull = new Float64Array(nSteps + 1);
  dfFull[0] = curve.discountFactor(0);
  for (let k = 0; k < nSteps; k++) dfFull[k + 1] = curve.discountFactor(times[k]);
  const fwdAtStep = new Float64Array(nSteps);
  for (let k = 0; k < nSteps; k++) {
    fwdAtStep[k] = -Math.log(dfFull[k + 1] / dfFull[k]) / dtYears;
  }

  const decay = Math.exp(-a * dtYears);
  const stdX = sigma * Math.sqrt((1.0 - Math.exp(-2.0 * a * dtYears)) / (2.0 * a));

  // Pre-allocate per-path Float64 arrays for X and r.
  const XPaths: Float64Array[] = new Array(nPaths);
  const rPaths: Float64Array[] = new Array(nPaths);
  for (let p = 0; p < nPaths; p++) {
    XPaths[p] = new Float64Array(nSteps);
    rPaths[p] = new Float64Array(nSteps);
  }

  const xCurr = new Float64Array(nPaths);
  const rUncorrCurr = new Float64Array(nPaths);
  const f0 = fwdAtStep[0];
  for (let p = 0; p < nPaths; p++) rUncorrCurr[p] = f0;
  const dPath = new Float64Array(nPaths);
  for (let p = 0; p < nPaths; p++) dPath[p] = 1.0;

  const dfSimulated = new Float64Array(nSteps);
  const rawErrorBps = new Float64Array(nSteps);
  const dfMarket = new Float64Array(nSteps);
  for (let k = 0; k < nSteps; k++) dfMarket[k] = curve.discountFactor(times[k]);

  for (let k = 0; k < nSteps; k++) {
    let sumDUncorr = 0;
    const dUncorrStep = new Float64Array(nPaths);

    for (let pair = 0; pair < nPairs; pair++) {
      const z = rng.nextNormal();
      const shock = stdX * z;

      // Path index for the +Z branch.
      const idx1 = pair * 2;
      const xNext1 = xCurr[idx1] * decay + shock;
      const rNext1 = fwdAtStep[k] + xNext1;
      dUncorrStep[idx1] = Math.exp(-0.5 * (rUncorrCurr[idx1] + rNext1) * dtYears);
      sumDUncorr += dPath[idx1] * dUncorrStep[idx1];
      xCurr[idx1] = xNext1;
      rUncorrCurr[idx1] = rNext1;

      // Antithetic -Z branch.
      const idx2 = pair * 2 + 1;
      const xNext2 = xCurr[idx2] * decay - shock;
      const rNext2 = fwdAtStep[k] + xNext2;
      dUncorrStep[idx2] = Math.exp(-0.5 * (rUncorrCurr[idx2] + rNext2) * dtYears);
      sumDUncorr += dPath[idx2] * dUncorrStep[idx2];
      xCurr[idx2] = xNext2;
      rUncorrCurr[idx2] = rNext2;
    }

    const eDUncorr = sumDUncorr / nPaths;
    rawErrorBps[k] = (eDUncorr - dfMarket[k]) * 1e4;

    let sumDCorr = 0;
    const correction = applyCorrection ? dfMarket[k] / eDUncorr : 1.0;
    for (let p = 0; p < nPaths; p++) {
      dPath[p] = dPath[p] * dUncorrStep[p] * correction;
      sumDCorr += dPath[p];
      XPaths[p][k] = xCurr[p];
      rPaths[p][k] = rUncorrCurr[p];
    }
    dfSimulated[k] = sumDCorr / nPaths;
  }

  return {
    seed: typeof seed === "bigint" ? seed : BigInt(seed),
    a,
    sigma,
    nPaths,
    horizonYears,
    dtYears,
    times,
    XPaths,
    rPaths,
    dfMarket,
    dfSimulated,
    rawErrorBps,
    corrected: applyCorrection,
  };
}

const ACT_360_ANNUAL = 365 / 360;

/**
 * Single-state HW forward/par-swap rate at time t for a latent factor X(t).
 *
 * Mirrors `projectHWToTenor` operation-for-operation for one (X, t, tenor) so
 * the reusable StochasticRatePath can reconstruct any tenor analytically
 * instead of averaging realized 1M spots. Kept separate from projectHWToTenor
 * (rather than refactoring it) to leave the verified research path bit-stable;
 * a unit test cross-checks numerical equality between the two.
 *
 * Convention (matches ZeroCurve.forwardSwapRate / projectHWToTenor):
 *   - τ < 1Y: simple-compounded forward.
 *   - τ ≥ 1Y, integer: par swap on annual ACT/360 fixed schedule.
 * P(t,T) = (P^M(0,T)/P^M(0,t)) · exp(-B(t,T)·X(t) - V(t,T)),
 *   V(t,T) = (σ²/4a)(1 - e^{-2at}) B(t,T)².
 */
export function hwForwardRate(
  X: number,
  a: number,
  sigma: number,
  curve: ZeroCurve,
  t: number,
  tenorYears: number,
): number {
  const tau = tenorYears;
  const expDecay = 1.0 - Math.exp(-2.0 * a * t);

  // Sub-1Y: simple-compounded forward.
  if (tau <= 1.0 + 1e-9) {
    const Btau = bFunction(a, tau);
    const ratio = curve.discountFactor(t + tau) / curve.discountFactor(t);
    const varTerm = ((sigma * sigma) / (4.0 * a)) * expDecay * Btau * Btau;
    const logP = Math.log(ratio) - Btau * X - varTerm;
    const Ptau = Math.exp(logP);
    return (1.0 / Ptau - 1.0) / tau;
  }

  // ≥ 1Y: forward par swap rate on annual ACT/360 schedule.
  const n = Math.round(tau);
  let annuity = 0;
  let Pn = 1.0;
  for (let k = 1; k <= n; k++) {
    const B = bFunction(a, k);
    const V = ((sigma * sigma) / (4.0 * a)) * expDecay * B * B;
    const logP = Math.log(curve.discountFactor(t + k) / curve.discountFactor(t)) - B * X - V;
    const Pk = Math.exp(logP);
    annuity += ACT_360_ANNUAL * Pk;
    Pn = Pk;
  }
  return (1.0 - Pn) / annuity;
}

/**
 * Project HW simulation paths to forward swap rate S(t, t+tau) for each
 * path/step via the analytical bond price formula.
 *
 * Convention (matches ZeroCurve.forwardSwapRate / bgmPricing.buildSwap):
 *   - τ < 1Y: simple-compounded forward (single-payment OIS).
 *   - τ ≥ 1Y, integer: par swap rate on annual ACT/360 fixed schedule:
 *       S(t, t+n) = (P(t,t) - P(t,t+n)) / Σ_k (365/360) · P(t, t+k)
 * Bond prices P(t, t+k) come from HW analytic formula:
 *   P(t,T) = (P^M(0,T)/P^M(0,t)) · exp(-B(t,T) · X(t) - V(t,T))
 *   V(t,T) = (σ²/4a)(1 - e^{-2at}) B(t,T)²
 */
export function projectHWToTenor(
  sim: HWSimulationResult,
  curve: ZeroCurve,
  tenorYears: number,
): Float64Array[] {
  const { a, sigma, times, XPaths, nPaths } = sim;
  const tau = tenorYears;
  const nSteps = times.length;

  // Sub-1Y: simple-compounded forward — original code path.
  if (tau <= 1.0 + 1e-9) {
    const Btau = bFunction(a, tau);
    const ratio = new Float64Array(nSteps);
    const varTerm = new Float64Array(nSteps);
    for (let k = 0; k < nSteps; k++) {
      const t = times[k];
      ratio[k] = curve.discountFactor(t + tau) / curve.discountFactor(t);
      varTerm[k] = ((sigma * sigma) / (4.0 * a)) * (1.0 - Math.exp(-2.0 * a * t)) * Btau * Btau;
    }
    const out: Float64Array[] = new Array(nPaths);
    for (let p = 0; p < nPaths; p++) {
      const path = new Float64Array(nSteps);
      for (let k = 0; k < nSteps; k++) {
        const logP = Math.log(ratio[k]) - Btau * XPaths[p][k] - varTerm[k];
        const Ptau = Math.exp(logP);
        path[k] = (1.0 / Ptau - 1.0) / tau;
      }
      out[p] = path;
    }
    return out;
  }

  // ≥ 1Y: forward par swap rate on annual schedule.
  const n = Math.round(tau);
  // Precompute B(τ_k) and per-step ratio/varTerm tables for each k=1..n.
  const Bk = new Float64Array(n + 1);
  for (let k = 1; k <= n; k++) Bk[k] = bFunction(a, k);
  const expDecay = new Float64Array(nSteps);
  for (let s = 0; s < nSteps; s++) {
    expDecay[s] = 1.0 - Math.exp(-2.0 * a * times[s]);
  }
  // ratioKS[k][s] = P^M(0, t_s + k) / P^M(0, t_s).
  const ratioKS: Float64Array[] = new Array(n + 1);
  for (let k = 1; k <= n; k++) {
    const r = new Float64Array(nSteps);
    for (let s = 0; s < nSteps; s++) {
      r[s] = curve.discountFactor(times[s] + k) / curve.discountFactor(times[s]);
    }
    ratioKS[k] = r;
  }
  const out: Float64Array[] = new Array(nPaths);
  for (let p = 0; p < nPaths; p++) {
    const path = new Float64Array(nSteps);
    for (let s = 0; s < nSteps; s++) {
      const X = XPaths[p][s];
      // Compute P(t, t+k) for k = 1..n; build annuity and tail bond.
      let annuity = 0;
      let Pn = 1.0;
      for (let k = 1; k <= n; k++) {
        const B = Bk[k];
        const V = ((sigma * sigma) / (4.0 * a)) * expDecay[s] * B * B;
        const logP = Math.log(ratioKS[k][s]) - B * X - V;
        const Pk = Math.exp(logP);
        annuity += ACT_360_ANNUAL * Pk;
        Pn = Pk;
      }
      // P(t, t) = 1, so numerator = 1 - Pn.
      path[s] = (1.0 - Pn) / annuity;
    }
    out[p] = path;
  }
  return out;
}

export interface PathDFResult {
  /** Dcorr[p][k] = terminally-corrected money-market DF to time (k+1)·dt on path p. */
  Dcorr: Float64Array[];
  /** max|E_p[D_p(t_k)] − P^M(0,t_k)| across horizons, in bp (martingale check). */
  meanErrBp: number;
}

/**
 * Reconstruct per-path money-market discount factors from the HW short-rate
 * paths, with a terminal multiplicative correction so the cross-path mean DF
 * equals the market DF at every horizon (E_p[D_p(t_k)] = P^M(0, t_k)). Verified
 * path-by-path identical to the simulator's own incremental correction; the
 * companion of `projectHWToTenor` for discounting rather than forward rates.
 *
 * Trapezoidal money-market accrual: D ∝ Π exp(−½(r_{k-1}+r_k)·dt), seeded at the
 * instantaneous t=0 forward f^M(0, dt). Used by stochastic FTP / OAS discounting.
 */
export function reconstructPathDF(sim: HWSimulationResult, curve: ZeroCurve): PathDFResult {
  const P = sim.nPaths;
  const nSteps = sim.times.length;
  const dt = sim.dtYears;
  const f0 = -Math.log(curve.discountFactor(dt)) / dt;

  const U: Float64Array[] = new Array(P);
  for (let p = 0; p < P; p++) {
    const u = new Float64Array(nSteps);
    let cum = 1;
    let rPrev = f0;
    for (let k = 0; k < nSteps; k++) {
      const rk = sim.rPaths[p][k];
      cum *= Math.exp(-0.5 * (rPrev + rk) * dt);
      u[k] = cum;
      rPrev = rk;
    }
    U[p] = u;
  }

  const corr = new Float64Array(nSteps);
  for (let k = 0; k < nSteps; k++) {
    let mean = 0;
    for (let p = 0; p < P; p++) mean += U[p][k];
    mean /= P;
    if (mean < 1e-15) throw new Error(`reconstructPathDF: DF mean collapsed to zero at step ${k}`);
    corr[k] = curve.discountFactor(sim.times[k]) / mean;
  }

  const Dcorr: Float64Array[] = new Array(P);
  for (let p = 0; p < P; p++) {
    const d = new Float64Array(nSteps);
    for (let k = 0; k < nSteps; k++) d[k] = U[p][k] * corr[k];
    Dcorr[p] = d;
  }

  let maxErr = 0;
  for (let k = 0; k < nSteps; k++) {
    let mean = 0;
    for (let p = 0; p < P; p++) mean += Dcorr[p][k];
    mean /= P;
    maxErr = Math.max(maxErr, Math.abs(mean - curve.discountFactor(sim.times[k])) * 1e4);
  }
  return { Dcorr, meanErrBp: maxErr };
}

File: ALM-Model-Lab/research/simulate_hw.py
"""Hull-White 1-factor Monte Carlo simulator.

Exact transition for the OU latent factor X under shifted-Gaussian HW:

    r(t) = f^M(0, t) + X(t),     dX = -a X dt + sigma dW.

Per simulation step:
    X_{t+dt} = X_t * exp(-a dt) + sigma * sqrt((1 - exp(-2 a dt)) / (2 a)) * Z
    r_uncorr_{t+dt} = f^M(0, t+dt) + X_{t+dt}
    D_step = exp(-0.5 (r_t + r_{t+dt}) dt)             # trapezoidal
    D_path *= D_step

Martingale correction (Glasserman Sec 4.5): scale every path's cumulative
discount factor at each step by  c_t = P^M(0, t) / E[D_uncorr],  yielding
an unbiased estimator of zero-coupon prices on the simulated paths.

This replaces the prototype's `runHullWhiteSimulation` (Gemini source lines
~530-660) with NumPy-vectorised path arithmetic, antithetic pairing, and
optional `apply_correction=False` for variance-reduction comparison runs.
"""

from __future__ import annotations

from dataclasses import dataclass
import os

import numpy as np

from bootstrap import ZeroCurve
from hw_pricing import b_function


@dataclass(frozen=True)
class HWSimulationResult:
    seed: int
    a: float
    sigma: float
    n_paths: int           # total paths (= 2 * n_pairs)
    horizon_years: float
    dt_years: float
    times: np.ndarray      # shape (n_steps,) simulation times t_1, ..., t_n
    X_paths: np.ndarray    # shape (n_paths, n_steps) latent factor X(t_k)
    r_paths: np.ndarray    # shape (n_paths, n_steps) short rate r(t_k) = f^M(0, t_k) + X
    df_market: np.ndarray  # shape (n_steps,) market DF(0, t_k)
    df_simulated: np.ndarray  # shape (n_steps,) simulated mean DF(0, t_k)
    raw_error_bps: np.ndarray  # uncorrected DF error (model minus market) in bps
    corrected: bool


def simulate_hw(
    curve: ZeroCurve,
    a: float,
    sigma: float,
    horizon_years: float = 30.0,
    dt_years: float = 1 / 12,
    n_pairs: int = 250,
    seed: int = 42,
    apply_correction: bool = True,
) -> HWSimulationResult:
    """Simulate HW1F paths under the spot-numeraire measure with antithetic pairing.

    Parameters
    ----------
    curve : ZeroCurve
        Bootstrapped initial zero curve.
    a, sigma : float
        Calibrated HW parameters.
    horizon_years : float
        Total simulation horizon.
    dt_years : float
        Step size (default monthly).
    n_pairs : int
        Number of antithetic pairs. Total paths = 2 * n_pairs.
    seed : int
        RNG seed for reproducibility.
    apply_correction : bool
        If True, apply multiplicative martingale correction at each step.

    Returns
    -------
    HWSimulationResult
        Latent factor and short-rate paths plus diagnostic arrays.
    """
    rng = np.random.default_rng(seed)
    n_steps = int(round(horizon_years / dt_years))
    times = np.arange(1, n_steps + 1) * dt_years
    n_paths = 2 * n_pairs

    # Forward rates derived from the curve, used as f^M(0, t) in r = f^M + X.
    # Using f(0, t) ~= -log(P(0, t+dt)/P(0, t))/dt at each grid point.
    grid_full = np.concatenate(([0.0], times))
    df_full = curve.discount_factor(grid_full)
    fwd_at_step = -np.log(df_full[1:] / df_full[:-1]) / dt_years  # f^M(0, t_k) for each step k

    # Exact OU transition coefficients.
    decay = np.exp(-a * dt_years)
    std_x = sigma * np.sqrt((1.0 - np.exp(-2.0 * a * dt_years)) / (2.0 * a))

    X = np.zeros((n_paths, n_steps), dtype=np.float64)
    r = np.zeros((n_paths, n_steps), dtype=np.float64)

    x_curr = np.zeros(n_paths, dtype=np.float64)
    # Initial r at t = 0 is the instantaneous forward at 0.
    f0 = float(fwd_at_step[0])
    r_uncorr_curr = np.full(n_paths, f0, dtype=np.float64)

    D_path = np.ones(n_paths, dtype=np.float64)
    df_simulated = np.empty(n_steps, dtype=np.float64)
    df_market = curve.discount_factor(times)
    raw_error_bps = np.empty(n_steps, dtype=np.float64)

    for k in range(n_steps):
        # Antithetic shocks: first half +Z, second half -Z.
        z = rng.standard_normal(n_pairs)
        shock = np.concatenate((std_x * z, -std_x * z))

        x_next = x_curr * decay + shock
        r_uncorr_next = fwd_at_step[k] + x_next

        D_step = np.exp(-0.5 * (r_uncorr_curr + r_uncorr_next) * dt_years)

        # Per-step uncorrected mean discount factor.
        e_d_uncorr = float(np.mean(D_path * D_step))
        raw_error_bps[k] = (e_d_uncorr - df_market[k]) * 1e4

        if apply_correction:
            correction = df_market[k] / e_d_uncorr
            D_path = D_path * D_step * correction
        else:
            D_path = D_path * D_step

        df_simulated[k] = float(np.mean(D_path))

        x_curr = x_next
        r_uncorr_curr = r_uncorr_next
        X[:, k] = x_curr
        r[:, k] = r_uncorr_next

    return HWSimulationResult(
        seed=seed,
        a=a,
        sigma=sigma,
        n_paths=n_paths,
        horizon_years=horizon_years,
        dt_years=dt_years,
        times=times,
        X_paths=X,
        r_paths=r,
        df_market=df_market,
        df_simulated=df_simulated,
        raw_error_bps=raw_error_bps,
        corrected=apply_correction,
    )


def project_hw_to_tenor(
    sim: HWSimulationResult,
    curve: ZeroCurve,
    tenor_years: float,
) -> np.ndarray:
    """Reconstruct simply-compounded forward F(t, t+tau) for each path/time.

    Uses the analytical HW bond price formula for P(t, t+tau) given X(t).
    Returns an array of shape (n_paths, n_steps).
    """
    a, sigma = sim.a, sim.sigma
    times = sim.times
    tau = tenor_years

    B_tau = float(b_function(a, tau))
    var_term = (sigma * sigma / (4.0 * a)) * (1.0 - np.exp(-2.0 * a * times)) * (B_tau ** 2)
    P0_t = curve.discount_factor(times)
    P0_T = curve.discount_factor(times + tau)
    ratio = P0_T / P0_t  # shape (n_steps,)

    # P(t, t+tau) = ratio * exp(-B(tau) X(t) - V(t, tau))
    log_P = np.log(ratio)[None, :] - B_tau * sim.X_paths - var_term[None, :]
    P_t_T = np.exp(log_P)
    return (1.0 / P_t_T - 1.0) / tau


if __name__ == "__main__":
    from pathlib import Path
    from market_data import load_market_snapshot
    from bootstrap import bootstrap_zero_curve
    from hw_calibrate import calibrate_hw

    snap = load_market_snapshot(
        Path(__file__).parent / "data" / os.environ.get("MARKET_JSON", "market_2025-09-30.json")
    )
    curve = bootstrap_zero_curve(snap)
    hw_fit = calibrate_hw(snap, curve)

    sim = simulate_hw(
        curve, hw_fit.a, hw_fit.sigma,
        horizon_years=30.0, dt_years=1 / 12, n_pairs=250, seed=20250930,
    )
    sim_no_corr = simulate_hw(
        curve, hw_fit.a, hw_fit.sigma,
        horizon_years=30.0, dt_years=1 / 12, n_pairs=250, seed=20250930,
        apply_correction=False,
    )

    print(f"HW simulation: a={hw_fit.a:.6f}, sigma={hw_fit.sigma:.6f}, "
          f"n_paths={sim.n_paths}, seed={sim.seed}")
    print(f"\n  step  t       DF_mkt    DF_corr   err_bp  DF_uncorr err_bp")
    for k in (0, 11, 59, 119, 239, 359):
        if k >= len(sim.times):
            continue
        print(
            f"  {k:4d}  {sim.times[k]:5.2f}  "
            f"{sim.df_market[k]:8.6f}  {sim.df_simulated[k]:8.6f}  "
            f"{(sim.df_simulated[k] - sim.df_market[k]) * 1e4:+7.3f}  "
            f"{sim_no_corr.df_simulated[k]:8.6f}  "
            f"{(sim_no_corr.df_simulated[k] - sim.df_market[k]) * 1e4:+7.3f}"
        )

    # 1Y forward path summary at a few horizons.
    fwd = project_hw_to_tenor(sim, curve, tenor_years=1.0)
    print(f"\n1Y forward rate distribution at selected horizons:")
    print(f"  horizon  mean    p5      p95")
    for k in (0, 11, 59, 119, 239):
        if k >= fwd.shape[1]:
            continue
        cs = fwd[:, k]
        print(
            f"  {sim.times[k]:6.2f}y  {cs.mean()*100:6.4f}%  "
            f"{np.percentile(cs, 5)*100:6.4f}%  "
            f"{np.percentile(cs, 95)*100:6.4f}%"
        )
